### PR TITLE
[spirv] support non-existing header file include

### LINKED ISSUE #3165: [SPIR-V] Line directives with filenames fail to compile for spirv-target with debug info
The following code fails to compile when targeting spirv + enabled debug information:

```HLSL
#line 1 "non_existing_file.txt"

struct PSInput
{
	float4 color : COLOR;
};

float4 PSMain(PSInput input) : SV_TARGET
{
	return input.color;
}
```

Command line:
```
dxc -E PSMain -T ps_6_0 -Zi -spirv test.hlsl
```

The output is:
```
dxc failed : error code 0x80004005.
```

We currently use paths to non-existing files in the line directives in our content pipeline. I assume that the content of the file shouldn't be read by the compiler.

It does compile fine without the spirv target.

### PR DESCRIPTION
**Pull request by jaebaek**
Fixes #3165

### CODE REVIEW

**Review by kuhar — APPROVED** [review]
Look fine to me. I'm not familiar with the code, but the change seem simple and passes both the public and internal tests.

File: tools/clang/lib/SPIRV/EmitVisitor.cpp [hunk 1]
```diff
@@ -521,14 +521,21 @@ bool EmitVisitor::visit(SpirvSource *inst) {
...
+      const auto &words = string::encodeSPIRVString(firstSnippet.getValue());
+      const auto numWordsInInstr = curInst.size() + words.size();
+      curInst[0] |= static_cast<uint32_t>(numWordsInInstr) << 16;
+      debugFileBinary.insert(debugFileBinary.end(), curInst.begin(),
+                             curInst.end());
+      debugFileBinary.insert(debugFileBinary.end(), words.begin(), words.end());
+    } else {
+      curInst[0] |= static_cast<uint32_t>(curInst.size()) << 16;
+      debugFileBinary.insert(debugFileBinary.end(), curInst.begin(),
+                             curInst.end());
```
Comment (dnovillo) [inline tools/clang/lib/SPIRV/EmitVisitor.cpp:537]: Just for my own education, do we really only realize if the file is missing when we attempt to read it?  If that's the case, should we factor out the `else` branch here with the `else` branch at line `539`?

**Comment by dnovillo** [comment]
> Fixes #3165

Could you also add a short description of how it fixes the issue?  It wasn't clear to me from the code itself (granted, I'm not too familiar with it).

File: tools/clang/lib/SPIRV/EmitVisitor.cpp (hunk 1 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/EmitVisitor.cpp:537]: Yeah, this `else` clause and the one on line 539 are the same and should be factored. Probably need to change line 515 to something like
```cpp
  if (spvOptions.debugInfoSource && inst->hasFile() && firstSnippet.hasValue()) {
```

File: tools/clang/unittests/SPIRV/CodeGenSpirvTest.cpp [hunk 2]
```diff
@@ -1559,6 +1559,9 @@ TEST_F(FileTest, SpirvLegalizationTextureBuffer) {
 TEST_F(FileTest, SpirvDebugOpSource) {
   runFileTest("spirv.debug.opsource.hlsl");
 }
+TEST_F(FileTest, SpirvDebugOpSourceNonExistingFile) {
+  runFileTest("spirv.debug.source.non.existing.file.hlsl");
```
Comment (ehsannas) [inline tools/clang/unittests/SPIRV/CodeGenSpirvTest.cpp:1563]: Better name would be `spirv.debug.source.non-existing-file.hlsl`

**Pull request merged**